### PR TITLE
Optimize SqlClient SNIPacket async paths

### PR DESCRIPTION
**Pull request by benaadams**
For `ReadFromStreamAsync(Stream stream, SNIAsyncCallback callback)`

`await stream.ReadAsync(new Memory<byte>(...))` 
is a more efficient path than 
`stream.ReadAsync(byte[], ...).ContinueWith(Task => ...)`

Move `WriteToStreamAsync(Stream stream, ...)` away from being `async void`

/cc @Wraith2 @saurabh500 @AfsanehR @keeratsingh @stephentoub

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs [hunk 1]
```diff
@@ -23,6 +23,7 @@ internal class SNIPacket : IDisposable, IEquatable<SNIPacket>
 
         private ArrayPool<byte>  _arrayPool = ArrayPool<byte>.Shared;
         private bool _isBufferFromArrayPool = false;
+        private ValueTask _asyncOperation;
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Why is this needed?  It appears to only ever be written to, not read.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs [hunk 2]
```diff
@@ -247,37 +249,50 @@ public void Reset()
         /// <param name="callback">Completion callback</param>
         public void ReadFromStreamAsync(Stream stream, SNIAsyncCallback callback)
         {
-            bool error = false;
-            
-            stream.ReadAsync(_data, 0, _capacity, CancellationToken.None).ContinueWith(t =>
+            // Treat local function as a static and pass all params otherwise as async will allocate
+            async ValueTask ReadFromStreamAsync(SNIPacket packet, SNIAsyncCallback cb, ValueTask<int> valueTask)
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: This can just return Task; returning ValueTask here doesn't add any benefit, and in fact incurs a little additional cost.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs [hunk 3]
```diff
@@ -247,37 +249,50 @@ public void Reset()
...
                 }
+            }
 
-                callback(this, error ? TdsEnums.SNI_ERROR : TdsEnums.SNI_SUCCESS);
-            },
-            CancellationToken.None,
-            TaskContinuationOptions.DenyChildAttach,
-            TaskScheduler.Default);
+            // Not complete or error call the async local function to complete
+            _asyncOperation = ReadFromStreamAsync(this, callback, vt);
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Where is this resulting task used?  Is it not possible for ReadFromStreamAsync to asynchronously throw?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs [hunk 4]
```diff
@@ -302,24 +317,47 @@ public void WriteToStream(Stream stream)
         /// Write data to a stream asynchronously
         /// </summary>
         /// <param name="stream">Stream to write to</param>
-        public async void WriteToStreamAsync(Stream stream, SNIAsyncCallback callback, SNIProviders provider, bool disposeAfterWriteAsync = false)
+        public void WriteToStreamAsync(Stream stream, SNIAsyncCallback callback, SNIProviders provider, bool disposeAfterWriteAsync = false)
         {
-            uint status = TdsEnums.SNI_SUCCESS;
-            try
+            // Treat local function as a static and pass all params otherwise as async will allocate
+            async ValueTask WriteToStreamAsync(SNIPacket packet, SNIAsyncCallback cb, SNIProviders providers, bool disposeAfter, ValueTask valueTask)
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: This should just return Task instead of ValueTask.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs [hunk 5]
```diff
@@ -302,24 +317,47 @@ public void WriteToStream(Stream stream)
...
+                    packet.Dispose();
+                }
             }
-            callback(this, status);
 
-            if (disposeAfterWriteAsync)
+            ValueTask vt = stream.WriteAsync(new Memory<byte>(_data, 0, _length), CancellationToken.None);
+
+            if (vt.IsCompletedSuccessfully)
             {
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: You should still call `.GetAwaiter().GetResult()` on `vt`; otherwise, the stream may not be able to reuse any `IValueTaskSource` used under the `ValueTask`, because it won't be able to know that the consumer is done with the instance.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 1 above)
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Was going to ask you about this 😄 

Is it necessary to hold a reference to prevent the Task potentially getting GC'd? Though I'd guess the Read/Write Op would be holding it?

Comment (stephentoub) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: > Is it necessary to hold a reference to prevent the Task potentially getting GC'd? Though I'd guess the Read/Write Op would be holding it?

Not if the async operation is implemented correctly: otherwise, how could it ever complete the task if it didn't keep it alive?

Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Sometimes I over think things 😉

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 3 above)
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Depends if we trust the callback (as the exception is caught then added to the callback)

I could recreate what .ContinueWith would do via
```csharp
try
{
    cb(packet, error ? TdsEnums.SNI_ERROR : TdsEnums.SNI_SUCCESS);
}
catch (Exception ex)
{
    // Create floating task to get picked up as an UnobservedTaskException
    Task.FromException(ex);
}
```
but that doesn't seem good?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 1 above)
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Done

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 2 above)
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Done

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 4 above)
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Done

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 5 above)
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Done

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs [hunk 6]
```diff
@@ -247,37 +247,50 @@ public void Reset()
...
 
-                callback(this, error ? TdsEnums.SNI_ERROR : TdsEnums.SNI_SUCCESS);
-            },
-            CancellationToken.None,
-            TaskContinuationOptions.DenyChildAttach,
-            TaskScheduler.Default);
+                cb(packet, error ? TdsEnums.SNI_ERROR : TdsEnums.SNI_SUCCESS);
+            }
+
+            ValueTask<int> vt = stream.ReadAsync(new Memory<byte>(_data, 0, _capacity), CancellationToken.None);
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Can you verify if my understanding of the intended behavior is correct?

The stream.ReadAsync posts a network read and if it successfully completes (synchronously) then the callback is invoked because `vt.IsCompletedSuccessfully` returned true. 
`vt.IsCompletedSuccessfully == false` means that network read is still pending or could have errored out and the client should wait for the valueTask to allocated a Task for completion.

Is this what you were intending? 
 
Can an error check be done on `vt`, before calling `ReadFromStreamAsync` to not allocate in case the stream read errors out? Or is it only possible to get the error by awaiting the ValueTask? 

IIUC, the optimization here is that if `stream.ReadAsync` completes synchronously, then we could be preventing Task allocation in those paths.

**Comment by saurabh500** [comment]
Looks like the netstandard targeted builds of SqlClient are failing. You may need to add the ValueTask specific code in separate files, or #if def them if the #ifdef is small. 

```
System\Data\SqlClient\SNI\SNIPacket.cs(251,83): error CS0246: The type or namespace name 'ValueTask<>' could not be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---0d2c9ac4\src\System.Data.SqlClient\src\System.Data.SqlClient.csproj]

```

File: src/System.Data.SqlClient/src/System.Data.SqlClient.csproj [hunk 7]
```diff
@@ -265,6 +265,12 @@
     </Compile>
     <Compile Include="System\Data\SqlClient\SqlFileStream.Windows.cs" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetsNetCoreApp)' == 'true'">
+    <Compile Include="System\Data\SqlClient\SNI\SNIPacket.NetCoreApp.cs" />
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System.Data.SqlClient.csproj]: This Compile Include could be moved after `    <Compile Include="System\Data\SqlClient\SqlConnectionStringBuilder.NetCoreApp.cs" />` instead of creating a new `ItemGroup`

File: src/System.Data.SqlClient/src/System.Data.SqlClient.csproj [hunk 8]
```diff
@@ -265,6 +265,12 @@
...
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetsNetCoreApp)' == 'false'">
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System.Data.SqlClient.csproj]: This should be '$(TargetsNetCoreApp)' != 'true'

The build system doesn't set an explicit `false` it simply doesn't set the value to true.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 6 above)
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: > Is this what you were intending?

Yes, though it should still allocate less than ContinueWith for the error path.

> Can an error check be done on vt, before calling ReadFromStreamAsync to not allocate in case the stream read errors out?

Yes, though would need to read `.Result` inside a try/catch, catching the exception to then add to the callback, so be doing extra logic and exception handling for the sync completion.

Is an error a common path vs successful completion? Is a Task much extra allocation on top of an exception with stacktrace? With exception handling being a petty slow path anyway is it worth optimizing the sync completion for it?

Comment (saurabh500) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: That's a fair argument. Now I understand why you did what you did 😃

File: src/System.Data.SqlClient/src/System.Data.SqlClient.csproj [hunk 9]
```diff
@@ -265,17 +265,15 @@
...
-    <Compile Include="System\Data\SqlClient\SNI\SNIPacket.NetStandard.cs" />
-  </ItemGroup>
   <ItemGroup Condition="'$(IsPartialFacadeAssembly)' != 'true' AND '$(OSGroup)' != 'AnyOS' AND '$(TargetsNetCoreApp)' == 'true'">
     <Compile Include="System\Data\Common\DbConnectionStringCommon.NetCoreApp.cs" />
     <Compile Include="System\Data\ProviderBase\DbConnectionPool.NetCoreApp.cs" />
     <Compile Include="System\Data\SqlClient\SqlConnectionString.NetCoreApp.cs" />
     <Compile Include="System\Data\SqlClient\SqlConnectionStringBuilder.NetCoreApp.cs" />
+    <Compile Include="System\Data\SqlClient\SNI\SNIPacket.NetCoreApp.cs" />
+  </ItemGroup>
+  <ItemGroup Condition="'$(IsPartialFacadeAssembly)' == 'true' OR '$(OSGroup)' == 'AnyOS' OR '$(TargetsNetCoreApp)' != 'true'">
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Data.SqlClient/src/System.Data.SqlClient.csproj]: Not sure packaging is my strong point :)

Should they be AND rather than OR also then?

[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Data.SqlClient/src/System.Data.SqlClient.csproj]: k, think I've got it

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs (hunk 3 above)
Comment (benaadams) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: It was `async void` in the Write path which is a hot knife through butter for an unhandled exception, so I imagine the callbacks are well behaved?

Comment (saurabh500) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs]: Yeah, callbacks are supposed to be well behaved. They seem to catch everything inside. 
About a year ago, I had done some analysis on these callbacks and came up with this conclusion.

**Comment by benaadams** [comment]
Seemed to do the trick; thank you for your help @saurabh500
